crates/ra_hir/src/db.rs

```diff
@@ -138,8 +144,12 @@ pub trait HirDatabase: DefDatabase {
     #[salsa::invoke(crate::ty::method_resolution::CrateImplBlocks::impls_in_crate_query)]
     fn impls_in_crate(&self, krate: Crate) -> Arc<CrateImplBlocks>;
 
-    #[salsa::invoke(crate::ty::traits::implements)]
-    fn implements(&self, trait_ref: TraitRef) -> Option<crate::ty::traits::Solution>;
+    #[salsa::invoke(crate::ty::traits::impls_for_trait)]
+    fn impls_for_trait(&self, krate: Crate, trait_: Trait) -> Arc<[ImplBlock]>;
+
+    #[salsa::invoke(crate::ty::traits::solver)]
+    #[salsa::volatile]
+    fn chalk_solver(&self, krate: Crate) -> Arc<Mutex<chalk_solve::Solver>>;
```

Comment:
`Arc<Mutex>` combo is pretty unexpected inside the hir. Am I correct that this is required because `Solver` does some caching on its own? So we use `volatile` query to not cache across source code modifications. It probably makes sense to put this into a comment.

I wonder what our long-term strategy here as well. Given that we want to kill type inference after every modifcation anyway, volatile query seems OK. 

OTOH, trait inference is only about top-level items, so it seems like it makes sense to cache this based on `CrateDefMap`

Reply:
Exactly. It's the same approach Chalk itself takes. And yeah, caching some trait checking results may make sense; we'll have to do quite a few queries for completion, especially if we want to provide results for all traits, whether in scope or not...

Reply:
I think this is the right approach. You **definitely** want the internal caching here, I think. That said, I think it's worth putting completions and things into their own sorts of queries, at least potentially -- so that they can be cached across edits.

Reply:
Sorry, what I wrote is a bit confusing -- and actually volatile isn't the *perfect* fit for this query I think, now that I think about it more. What I mean is that I think it would be nice if we can figure out that a given edit didn't (say) disturb the set of impls, and hence it doesn't affect the completions. But volatile is a bit too coarse for that. I'm not sure the best way to handle it in salsa today, it'd probably require some sort of extension. volatile is what I think we should do to start. :)

Reply:
>I'm not sure the best way to handle it in salsa today, it'd probably require some sort of extension

A hacky solution would be the followoing:

```rust

trait HirDb {
    // a usual, cached salsa query
    fn chalk_solver(&self krate: Crate);
}

fn chalk_solver(db: &impl HirDb, krate: Crate) {
    let _ = db.all_impls(krate); // dummy read to register dependency edge
    Arc::new(Mutex::new(ChalkSolver::new()))
}
```

Reply:
That would have to collect all impls unnecessarily though...